tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/XUnit/XUnitTestRunnerRunAsyncIntegration.cs

```diff
@@ -121,77 +129,100 @@ internal static async Task<TReturn> OnAsyncMethodEnd<TTarget, TReturn>(TTarget i
+            if (messageBus is { TestIsNew: true, AbortByThreshold: false } or { FlakyRetryEnabled: true }
+             && returnValue.TryDuckCast<IRunSummary>(out var runSummary))
             {
+                var isRetryEnabled = messageBus.FlakyRetryEnabled;
                 var index = messageBus.ExecutionIndex;
                 if (index == 0)
                 {
-                    // Let's make decisions based on the first execution regarding slow tests
-                    var duration = TraceClock.Instance.UtcNow - testRunnerState.StartTime;
-                    messageBus.TotalExecutions = Common.GetNumberOfExecutionsForDuration(duration);
+                    // Let's make decisions based on the first execution regarding slow tests or retry failed test feature
+                    if (isRetryEnabled)
+                    {
+                        messageBus.TotalExecutions = CIVisibility.Settings.FlakyRetryCount + 1;
+                    }
+                    else
+                    {
+                        var duration = TraceClock.Instance.UtcNow - testRunnerState.StartTime;
+                        messageBus.TotalExecutions = Common.GetNumberOfExecutionsForDuration(duration);
+                    }
+
                     messageBus.ExecutionNumber = messageBus.TotalExecutions - 1;
                 }
 
                 if (messageBus.ExecutionNumber > 0)
                 {
-                    var retryNumber = messageBus.ExecutionIndex + 1;
-                    // Set the retry as a continuation of this execution. This will be executing recursively until the execution count is 0/
-                    Common.Log.Debug<int, int>("EFD: [Retry {Num}] Test class runner is duck casted, running a retry. [Current retry value is {Value}]", retryNumber, messageBus.ExecutionNumber);
-                    var innerReturnValue = await ((Task<TReturn>)testRunnerState.TestRunner.RunAsync()).ConfigureAwait(false);
-                    if (innerReturnValue.TryDuckCast<IRunSummary>(out var innerRunSummary) &&
-                        returnValue.TryDuckCast<IRunSummary>(out var runSummary))
+                    if (isRetryEnabled && (Interlocked.Decrement(ref _totalRetries) < 1 || runSummary.Failed == 0))
                     {
-                        Common.Log.Debug<int>("EFD: [Retry {Num}] Aggregating results.", retryNumber);
-                        runSummary.Aggregate(innerRunSummary);
+                        if (runSummary.Failed == 0)
+                        {
+                            Common.Log.Debug("EFD/Retry: [FlakyRetryEnabled] A non failed test execution was detected, skipping the remaining executions.");
+                        }
+                        else
+                        {
+                            Common.Log.Debug<int>("EFD/Retry: [FlakyRetryEnabled] Exceeded de number of total retries. [{Number}]", CIVisibility.Settings.TotalFlakyRetryCount);
+                        }
```

Comment:
tiny nit: I would probably avoid duplicating the conditions here as I think it makes it easier to follow

```csharp
                    if (isRetryEnabled)
                    {
                        var remainingRetries = Interlocked.Decrement(ref _totalRetries);
                        if (runSummary.Failed == 0)
                        {
                            Common.Log.Debug("EFD/Retry: [FlakyRetryEnabled] A non failed test execution was detected, skipping the remaining executions.");
                        }
                        else if (remainingRetries < 1)
                        {
                            Common.Log.Debug<int>("EFD/Retry: [FlakyRetryEnabled] Exceeded number of total retries. [{Number}]", CIVisibility.Settings.TotalFlakyRetryCount);
                        }
```

Reply:
done in [957e5d0](https://github.com/DataDog/dd-trace-dotnet/pull/6061/commits/957e5d062551cc63f7e537ded6dd73d69b8c0933)